### PR TITLE
[threads] ref.i31_shared

### PR DESCRIPTION
**Pull request by tlively**
Implement `ref.i31_shared` the new instruction for creating references
to shared i31s. Implement binary and text parsing and emitting as well
as interpretation. Copy the upstream spec test for i31 and modify it so
that all the heap types are shared. Comment out some parts that we do
not yet support.

### CODE REVIEW

**Comment by tlively** [comment]
The test failures here and in #6736 will be resolved by #6737.

File: src/wasm.h [hunk 1]
```diff
@@ -1514,6 +1514,7 @@ class RefI31 : public SpecificExpression<Expression::RefI31Id> {
   RefI31() = default;
   RefI31(MixedArena& allocator) {}
 
+  Shareability share;
```
Comment (kripken) [inline src/wasm.h:1517]: Can this not be part of the `type`? We don't seem to have such a field on anything else.

Comment (tlively) [reply src/wasm.h:1517]: Hmm, I guess it could.

Comment (tlively) [reply src/wasm.h]: See the last commmit for this change.

File: src/passes/Print.cpp [hunk 2]
```diff
@@ -2076,7 +2076,10 @@ struct PrintExpressionContents
     o << std::max(curr->tuple->type.size(), size_t(2)) << " ";
     o << curr->index;
   }
-  void visitRefI31(RefI31* curr) { printMedium(o, "ref.i31"); }
+  void visitRefI31(RefI31* curr) {
+    printMedium(
+      o, curr->type.getHeapType().isShared() ? "ref.i31_shared" : "ref.i31");
```
Comment (kripken) [inline src/passes/Print.cpp:2081]: I think you may need to guard here against `curr->type` being unreachable (when the i32 child is unreachable), in which case `getHeapType` errors.

**Review by kripken — APPROVED** [review]

**Pull request merged**